### ISSUE TITLE
Default settings for parallelism and queue type causing performance degradation

### ISSUE THREAD

**Issue by gab-satchi** [knative-sandbox/eventing-rabbitmq#835]
**Describe the bug**
The default settings of a Broker/Trigger setup uses:
- Parallelism set to 1. This prefetches 1 message at a time and processes it using a single worker
- Queue type set to quorum. 

With quorum queues, the acks can take a while to process. The acks need to be written to disk and propagated to all the nodes in the cluster. This causes a bottle-neck where the consumer is "starving" when the consumer is fetching a single message at a time. 

Options:
- Set default queue type to `classic` since it's configurable and user can modify.
- Set a > 1 default for parallelism.
- Split parallelism into `prefetchCount` and `workers`.
  - option 1: prefetchCount isn't configurable but parallelism becomes number of workers. Prefetch is set to a static value like 100.
  - option 2: prefetchCount and workers can both be configurable. Can still default prefetch to be > 1 while allowing it to be configurable by users.

 **Knative release version**
upcoming release

 **Additional context**
The following are two runs of the same test. A producer producing 1000 messages a second and a single consumer consuming them. The key difference between the two is the queue type. We can see the drop in throughput when using Quorum queues with prefetch set to 1.

*Note:* This issue is only present with a very low prefetch number. The performance is pretty similar when running the same tests with parallelism set to 100.

**Classic Queue**
![parallel-1-throughput](https://user-images.githubusercontent.com/4602447/175541013-a81493d8-adf1-41c6-9539-e9b683629bcb.png)

The throughput is around 500-600 until the producer finishes and then jumps to well over 1000.

**Quorum Queue**
![parallel-1-throughput](https://user-images.githubusercontent.com/4602447/175541192-498f6293-8815-446d-a3bc-7d7446d14985.png)

The throughput barely breaks 400 even after the producer is finished.

**Comment by evankanderson** [comment]
Would it make sense to default `parallelism` to a value higher than 1 by default, and document the performance tradeoffs with `parallelism=1`?

**Comment by gab-satchi** [comment]
> Just to clarify, is the performance penalty mainly due to quorum vs classic or do we have two independent issues with the queue type and parallelism settings?

Just the one issue. When using classic queues, we do not experience the penalty even when setting parallelism=1

> Would it make sense to default parallelism to a value higher than 1 by default, and document the performance tradeoffs with parallelism=1?

We lose ordering when setting parallelism to > 1 which makes it more than a performance tradeoff. My suggestion would be to do: 

```
option 1: prefetchCount isn't configurable but parallelism becomes number of workers. Prefetch is set to a static value like 100.
``` 

Ordering would be maintained but our dispatcher would fetch 100 messages at a time and do a batch ack.

**Comment by evankanderson** [comment]
The Broker interface explicitly does not guarantee ordering. 😁 

I understand that ordering is useful in some cases, but I worry about having our default be "perform poorly until configured otherwise".

**Comment by gab-satchi** [comment]
Added increasing the default `parallelism` as another option. It's another easy fix compared to implementing batch acks.

**Comment by embano1** [comment]
> The Broker interface explicitly does not guarantee ordering. 😁
> 
> I understand that ordering is useful in some cases, but I worry about having our default be "perform poorly until configured otherwise".

When we added parallelism we had this discussion of keeping the default (FIFO) or increasing performance (throughput). We decided to keep FIFO (parallelism: 1) because that was the behavior of existing deployments, i.e. we‘d potentially break RMQ Knative users ([Hyrum‘s law](https://www.hyrumslaw.com) in action). 

I understand that ordering is not guaranteed by the SPEC but users rely on what they get which in this case is FIFO semantics by default in RMQ Knative. Especially in certain VEBA use cases we‘ve seen users wanting (assuming) FIFO for correctness.

My stance is make it „safe“ by default, provide knobs to make it fast. And this is what I see from @gab-satchi‘s first proposal (option 1) with bumping up prefetch, while defaulting concurrency to 1 (for quorum queues).

**Comment by xtreme-sameer-vohra** [comment]
> option 1: prefetchCount isn't configurable but parallelism becomes number of workers. Prefetch is set to a static value like 100.

- Would a user loose ordering in this scenario ? The batch fetched from RMQ would be in order but would using a worker pool allow guarantee of order ?
- Are there message delivery tradeoffs that would be made with option 1 since it would be a batch `ack` ?

**Comment by gab-satchi** [comment]
> Would a user loose ordering in this scenario ? The batch fetched from RMQ would be in order but would using a worker pool allow guarantee of order ?

If there's more than 1 worker then the user will lose ordering. If prefetch is set to 100 but a single worker is consuming messages (parallelism=1) then order will be maintained.

> Are there message delivery tradeoffs that would be made with option 1 since it would be a batch ack ?

With prefetch > 1, we aren't forced to do batch ack. We can still ack 1 at a time and maybe can do it asynchronously so the next message to process isn't waiting for the ack to complete.

**Comment by gab-satchi** [comment]
I did a spike with the following setup: 
`Prefetch set to 100 but a single worker consumes the messages` 

FIFO was maintained and when running the same performance test, the performance improved significantly. Receiver throughput basically matched the sender (similar to parallelism set to 100 test results):
![parallel-1-throughput](https://user-images.githubusercontent.com/4602447/177631901-387042ef-a479-42d1-a287-b4e6d4e9217f.png)

@xtreme-sameer-vohra, there were some concerns around maintaining at-least-once delivery if prefetch > 1. I confirmed that we can continue to ACK a single message at a time. Did a sanity test where the dispatcher was dropping about 25% of the acks and I continue to see those messages Unacked in the queue.
